### ISSUE TITLE
Cold storage support for partitions

### ISSUE THREAD

**Issue by bluestreak01** [questdb/roadmap#5]
### Summary

Chronologically older partitions should go to a cold storage to save cost. At the same time data in these partitions should remain accessible, although with performance penalty. There should be both manual and automated method to "mark" partition or group of partitions as cold.

### Intended outcome

Partition is "marked" as cold. The physical location of data in this partition will change in read-consistent fashion. Data in partition remains available, row counts remains consistent with data being there. When accessed, data is delivered to a query with lag, consistent with fetching data from cold storage.

**Comment by snenkov** [comment]
I am curious how this exactly would work. I understand the high-level logic, but I don't see how this related to QDB and what QDB would do to assist this i.e. how would QDB covert data to cold-storage to save cost? Isn't that something that's depended on the unique setup of each server and not on QDB?

**Comment by bluestreak01** [comment]
Cold storage for all intents and purposes is a device that does not support any useful form of random access. Our intent here is to implement a virtual area backed by local storage that would intelligently swap data from and to cold storage on access. Ultimately there will be addressable partition in table that would incur a time penalty on access if data happens to be remote.

**Comment by snenkov** [comment]
Okay, interesting. Thank, Vlad. I am starting to use QDB in a similar fashion, hence why I was wondering what a natural QDB solution would be like.

**Comment by postol** [comment]
Looking forward to this feature - for use cases where you need to store large amount of data this is the most important feature one needs. One additional comment to data swapping - there will have to be some staging area with expiry and size limits for swapped data otherwise you would blow up the local storage very quickly if you start querying too much data living in cold storage. Also there should be sanity checks that if someone asks for data in cold storage that would be too large (won't fit to staging area) such query will be rejected or similar approach...

**Comment by jjuraszek** [comment]
ideal would be to store cold on some external service like s3. sufficient would be defining separate file path on the server. is this in the scope of this feature?

**Comment by puzpuzpuz** [comment]
> ideal would be to store cold on some external service like s3. sufficient would be defining separate file path on the server. is this in the scope of this feature?

Yes, cold partitions will be stored on S3 or any other BLOB storage supported by [OpenDAL](https://opendal.apache.org/).

**Comment by tisonkun** [comment]
@puzpuzpuz May I ask where the development of integrations with OpenDAL happens?

**Comment by puzpuzpuz** [comment]
> @puzpuzpuz May I ask where the development of integrations with OpenDAL happens?

It's a part of QuestDB Enterprise.

**Comment by tisonkun** [comment]
> > @puzpuzpuz May I ask where the development of integrations with OpenDAL happens?
> 
> It's a part of QuestDB Enterprise.

Good to know. I'm maintaining the OpenDAL Java binding. Recently, we implemented Java IO's abstractions to support streaming read/write - https://github.com/apache/opendal/pull/4626

I hope the OpenDAL integration a basic non-competing part of your Enterprise edition so that we can deal with the challenges together in the upstream.

I remember QuestDB members reported issues on the distribution, and I suppose that there should be still other features missing to make a complete Java SDK. Looking forward to your inputs :D

**Comment by puzpuzpuz** [comment]
Hi @tisonkun,

We're using OpenDAL from Rust, so Java bindings aren't relevant for us.
